File: web/src/artifacts.ts
import type { Artifact, MessagePart } from "./types";

const ARTIFACT_BLOCK_RE = /```\s*questrag-artifact\s*([\s\S]*?)```/g;
const ARTIFACT_FENCE_START_RE = /```\s*questrag-artifact\s*/g;
const ARTIFACT_FENCE_PREFIX = "```questrag-artifact";

export function parseMessageParts(raw: string): MessagePart[] {
  const parts: MessagePart[] = [];
  let cursor = 0;

  for (const match of raw.matchAll(ARTIFACT_BLOCK_RE)) {
    const index = match.index ?? 0;
    const before = raw.slice(cursor, index);
    if (before) {
      parts.push({ type: "markdown", content: before });
    }

    const artifact = parseArtifact(match[1]);
    if (artifact?.type === "chart") {
      parts.push({ type: "chart", artifact });
    } else if (artifact?.type === "report") {
      parts.push({ type: "report", artifact });
    } else if (artifact?.type === "application") {
      parts.push({ type: "application", artifact });
    } else {
      parts.push({ type: "markdown", content: match[0] });
    }

    cursor = index + match[0].length;
  }

  const tail = raw.slice(cursor);
  if (tail) {
    parts.push({ type: "markdown", content: tail });
  }

  return mergeMarkdownParts(parts);
}

export function parseStreamingMessageParts(raw: string): { parts: MessagePart[]; bufferingArtifact: boolean } {
  const openFenceIndex = findOpenArtifactFence(raw);
  if (openFenceIndex < 0) {
    const prefixIndex = findTrailingArtifactFencePrefix(raw);
    if (prefixIndex >= 0) {
      return {
        parts: parseMessageParts(raw.slice(0, prefixIndex)),
        bufferingArtifact: true,
      };
    }
    return { parts: parseMessageParts(raw), bufferingArtifact: false };
  }
  return {
    parts: parseMessageParts(raw.slice(0, openFenceIndex)),
    bufferingArtifact: true,
  };
}

function findTrailingArtifactFencePrefix(raw: string) {
  const normalized = (value: string) => value.replace(/\s+/g, "");
  const maxLength = Math.min(raw.length, ARTIFACT_FENCE_PREFIX.length + 8);
  for (let length = maxLength; length >= 3; length -= 1) {
    const start = raw.length - length;
    const suffix = raw.slice(start);
    const compact = normalized(suffix).toLowerCase();
    if (ARTIFACT_FENCE_PREFIX.startsWith(compact) && isOpeningFencePosition(raw, start)) {
      return start;
    }
  }
  return -1;
}

function isOpeningFencePosition(raw: string, index: number) {
  const before = raw.slice(0, index);
  const fenceCount = before.match(/```/g)?.length || 0;
  return fenceCount % 2 === 0;
}

function findOpenArtifactFence(raw: string) {
  ARTIFACT_FENCE_START_RE.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = ARTIFACT_FENCE_START_RE.exec(raw)) !== null) {
    const closeIndex = raw.indexOf("```", ARTIFACT_FENCE_START_RE.lastIndex);
    if (closeIndex < 0) return match.index;
    ARTIFACT_FENCE_START_RE.lastIndex = closeIndex + 3;
  }
  return -1;
}

function parseArtifact(value: string): Artifact | null {
  try {
    const parsed = JSON.parse(value.trim());
    if (parsed?.type === "chart" && Array.isArray(parsed.data)) {
      return parsed;
    }
    if (parsed?.type === "report" && typeof parsed.content === "string") {
      return parsed;
    }
    if (parsed?.type === "application" && typeof parsed.case_id === "string" && typeof parsed.title === "string") {
      return parsed;
    }
  } catch {
    return null;
  }
  return null;
}

function mergeMarkdownParts(parts: MessagePart[]): MessagePart[] {
  const merged: MessagePart[] = [];
  for (const part of parts) {
    const last = merged[merged.length - 1];
    if (part.type === "markdown" && last?.type === "markdown") {
      last.content += part.content;
    } else {
      merged.push(part);
    }
  }
  return merged.length ? merged : [{ type: "markdown", content: "" }];
}

export function appendMarkdownToParts(parts: MessagePart[], text: string): MessagePart[] {
  const next = [...parts];
  const last = next[next.length - 1];
  if (last?.type === "markdown") {
    next[next.length - 1] = { ...last, content: last.content + text };
  } else {
    next.push({ type: "markdown", content: text });
  }
  return next;
}

File: quest_rag/chat_memory/artifacts.py
import json
import re
from typing import Any

from quest_rag.schemas.schemas import ChartArtifact

ARTIFACT_BLOCK_RE = re.compile(r"```\s*questrag-artifact\s*([\s\S]*?)```")


def parse_message_parts(raw: str) -> list[dict[str, Any]]:
    parts: list[dict[str, Any]] = []
    cursor = 0

    for match in ARTIFACT_BLOCK_RE.finditer(raw or ""):
        before = raw[cursor:match.start()]
        if before:
            parts.append({"type": "markdown", "content": before})

        artifact = parse_artifact(match.group(1))
        if artifact and artifact.get("type") in {"chart", "report", "application"}:
            parts.append({"type": artifact["type"], "artifact": artifact})
        else:
            parts.append({"type": "markdown", "content": match.group(0)})

        cursor = match.end()

    tail = (raw or "")[cursor:]
    if tail:
        parts.append({"type": "markdown", "content": tail})

    return merge_markdown_parts(parts)


def normalize_message_parts(raw: str, parts: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
    if not raw:
        return parts or []
    if not parts:
        return parse_message_parts(raw)
    if len(parts) == 1 and parts[0].get("type") == "markdown" and ARTIFACT_BLOCK_RE.search(raw):
        return parse_message_parts(raw)
    return parts


def parse_artifact(value: str) -> dict[str, Any] | None:
    try:
        parsed = json.loads(value.strip())
    except json.JSONDecodeError:
        return None

    if parsed.get("type") == "chart" and isinstance(parsed.get("data"), list):
        try:
            return _dump_model(ChartArtifact(**parsed))
        except Exception:
            return None
    if parsed.get("type") == "report" and isinstance(parsed.get("content"), str):
        return parsed
    if (
        parsed.get("type") == "application"
        and isinstance(parsed.get("case_id"), str)
        and isinstance(parsed.get("title"), str)
    ):
        return parsed
    return None


def merge_markdown_parts(parts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    merged: list[dict[str, Any]] = []
    for part in parts:
        last = merged[-1] if merged else None
        if part.get("type") == "markdown" and last and last.get("type") == "markdown":
            last["content"] = f"{last.get('content', '')}{part.get('content', '')}"
        else:
            merged.append(part)
    return merged or [{"type": "markdown", "content": ""}]


def _dump_model(model) -> dict[str, Any]:
    if hasattr(model, "model_dump"):
        return model.model_dump(mode="json")
    return model.dict()
